File: ios/SiliconBuddy/Reach/IntentMapping.swift
import Foundation

/// The part of "Hey Siri, load Gemma on my Mac" that is worth testing.
///
/// An intent parameter is a phrase a person said or typed, and the Mac wants an id. The
/// mapping between the two is where a shortcut either works or quietly loads the wrong
/// model, so it is here, in one place, with no Siri anywhere near it.
public enum IntentMapping {

    /// Why an intent could not be carried out, in words that Siri can read aloud.
    public enum Refusal: Error, Equatable, LocalizedError {
        case notPaired
        case notAllowed
        case emptyPrompt
        case noSuchModel(String)
        case ambiguous(String, [String])

        public var errorDescription: String? {
            switch self {
            case .notPaired:
                "Silicon Buddy isn't paired with a Mac yet. Open the app and pair first."
            case .notAllowed:
                BuddyAPI.DeviceScope.chat.explanation
            case .emptyPrompt:
                "There was nothing to ask."
            case .noSuchModel(let name):
                "No model on your Mac is called \"\(name)\"."
            case .ambiguous(let name, let candidates):
                "\"\(name)\" matches \(candidates.count) models: "
                    + candidates.prefix(3).joined(separator: ", ")
                    + ". Say more of the name."
            }
        }
    }

    /// How much an intent may generate. Smaller than the chat screen's on purpose: a
    /// spoken answer that runs for two thousand tokens is not an answer, it is a
    /// monologue, and Siri cuts it off anyway.
    public static let spokenMaxTokens = 512

    /// "Ask my Mac <text>" — one question, no history. A Shortcuts action has no
    /// transcript behind it, and pretending otherwise would attach whatever the phone
    /// last talked about to an unrelated question.
    public static func chatRequest(
        prompt: String, images: [String] = [], maxTokens: Int = spokenMaxTokens
    ) throws -> ControlAPI.ChatRequest {
        let text = prompt.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !text.isEmpty || !images.isEmpty else { throw Refusal.emptyPrompt }
        return ControlAPI.ChatRequest(
            messages: [ControlAPI.ChatRequest.Message(role: "user", content: text, images: images)],
            maxTokens: maxTokens
        )
    }

    /// "Load <model>" — a phrase, against what the Mac actually has installed.
    ///
    /// Exact id first, then exact name, then a unique prefix, then a unique containment.
    /// Each step is narrower than the next, and a step that matches more than one model
    /// refuses rather than picking: loading the wrong 27B model costs minutes and a
    /// gigabyte of memory, and the person is standing there listening.
    public static func resolveModel(
        named phrase: String, in installed: [ControlAPI.InstalledModel], scope: BuddyAPI.DeviceScope
    ) throws -> ControlAPI.InstalledModel {
        guard scope.canControl else { throw Refusal.notAllowed }
        let needle = normalise(phrase)
        guard !needle.isEmpty else { throw Refusal.noSuchModel(phrase) }

        if let exact = installed.first(where: { normalise($0.id) == needle }) { return exact }

        let byName = installed.filter { normalise($0.name) == needle }
        if byName.count == 1 { return byName[0] }
        if byName.count > 1 { throw Refusal.ambiguous(phrase, byName.map(\.name)) }

        // The bare model id without its quantization — "gemma-3-12b-it" for
        // "gemma-3-12b-it@Q4_K_M" — which is how a person says it and how the catalog
        // prints it.
        let byBaseID = installed.filter {
            normalise($0.id.split(separator: "@").first.map(String.init) ?? $0.id) == needle
        }
        if byBaseID.count == 1 { return byBaseID[0] }
        if byBaseID.count > 1 { throw Refusal.ambiguous(phrase, byBaseID.map(\.name)) }

        let byPrefix = installed.filter { normalise($0.name).hasPrefix(needle) }
        if byPrefix.count == 1 { return byPrefix[0] }
        if byPrefix.count > 1 { throw Refusal.ambiguous(phrase, byPrefix.map(\.name)) }

        let byContains = installed.filter {
            normalise($0.name).contains(needle) || normalise($0.id).contains(needle)
        }
        if byContains.count == 1 { return byContains[0] }
        if byContains.count > 1 { throw Refusal.ambiguous(phrase, byContains.map(\.name)) }

        throw Refusal.noSuchModel(phrase)
    }

    public static func loadRequest(
        named phrase: String, in installed: [ControlAPI.InstalledModel], scope: BuddyAPI.DeviceScope
    ) throws -> ControlAPI.LoadRequest {
        // An installed model carries its quantization in its id, so the request needs
        // nothing else: the Mac already knows what it has on disk.
        ControlAPI.LoadRequest(modelID: try resolveModel(named: phrase, in: installed, scope: scope).id)
    }

    /// "What is loaded on my Mac" — one sentence, spoken.
    public static func loadedSummary(_ status: ControlAPI.Status, macName: String?) -> String {
        let mac = macName ?? "your Mac"
        guard let name = status.loadedModelName ?? status.loadedModelID else {
            return "Nothing is loaded on \(mac) right now."
        }
        var sentence = "\(mac) has \(name) loaded"
        if let context = status.contextLength, context > 0 {
            sentence += ", with a \(contextPhrase(context)) context"
        }
        if let rate = status.lastGenerationTokensPerSecond, rate > 0 {
            sentence += String(format: ", last answering at %.0f tokens a second", rate)
        }
        return sentence + "."
    }

    /// "65,536" spoken as "64K", because that is how anybody says it.
    static func contextPhrase(_ tokens: Int) -> String {
        guard tokens >= 1024 else { return "\(tokens) token" }
        let thousands = Double(tokens) / 1024
        return thousands == thousands.rounded()
            ? "\(Int(thousands))K" : String(format: "%.1fK", thousands)
    }

    /// Case, spacing and punctuation are not how a person distinguishes two models.
    static func normalise(_ text: String) -> String {
        text.lowercased()
            .replacingOccurrences(of: "_", with: "-")
            .replacingOccurrences(of: " ", with: "-")
            .trimmingCharacters(in: CharacterSet(charactersIn: "-.,!? "))
    }
}

File: android/app/src/main/java/dev/siliconoptimizer/buddy/reach/IntentMapping.kt
package dev.siliconoptimizer.buddy.reach

import dev.siliconoptimizer.buddy.transport.ChatMessageWire
import dev.siliconoptimizer.buddy.transport.ChatRequest
import dev.siliconoptimizer.buddy.transport.DeviceScope
import dev.siliconoptimizer.buddy.transport.InstalledModel
import dev.siliconoptimizer.buddy.transport.LoadRequest
import dev.siliconoptimizer.buddy.transport.Status

/**
 * The part of "Hey Google, ask my Mac…" that is worth testing.
 *
 * A shortcut parameter is a phrase a person said or typed, and the Mac wants an id. The
 * mapping between the two is where a shortcut either works or quietly loads the wrong
 * model, so it is here, in one place, with no Assistant anywhere near it.
 */
object IntentMapping {

    /** Why a shortcut could not be carried out, in words that can be read aloud. */
    sealed class Refusal(message: String) : Exception(message) {
        object NotPaired : Refusal(
            "Silicon Buddy isn't paired with a Mac yet. Open the app and pair first.",
        )

        object NotAllowed : Refusal(
            "This device is paired for chat only. Loading, installing and rendering are " +
                "hidden. Pair it again with full control from Settings, Silicon Buddy on the Mac.",
        )

        object EmptyPrompt : Refusal("There was nothing to ask.")

        class NoSuchModel(val phrase: String) :
            Refusal("No model on your Mac is called \"$phrase\".")

        class Ambiguous(val phrase: String, val candidates: List<String>) : Refusal(
            "\"$phrase\" matches ${candidates.size} models: " +
                candidates.take(3).joinToString(", ") + ". Say more of the name.",
        )
    }

    /**
     * How much a shortcut may generate. Smaller than the chat screen's on purpose: a
     * spoken answer that runs for two thousand tokens is not an answer, it is a
     * monologue, and the Assistant cuts it off anyway.
     */
    const val SPOKEN_MAX_TOKENS = 512

    /**
     * "Ask my Mac <text>" — one question, no history. A shortcut has no transcript
     * behind it, and pretending otherwise would attach whatever the phone last talked
     * about to an unrelated question.
     */
    fun chatRequest(
        prompt: String,
        images: List<String> = emptyList(),
        maxTokens: Int = SPOKEN_MAX_TOKENS,
    ): ChatRequest {
        val text = prompt.trim()
        if (text.isEmpty() && images.isEmpty()) throw Refusal.EmptyPrompt
        return ChatRequest(
            messages = listOf(ChatMessageWire("user", text, images)),
            maxTokens = maxTokens,
        )
    }

    /**
     * "Load <model>" — a phrase, against what the Mac actually has installed.
     *
     * Exact id first, then exact name, then the id without its quantization, then a
     * unique prefix, then a unique containment. Each step is narrower than the next,
     * and a step that matches more than one model refuses rather than picking: loading
     * the wrong 27B costs minutes and a gigabyte of memory, and the person is standing
     * there listening.
     */
    fun resolveModel(
        phrase: String,
        installed: List<InstalledModel>,
        scope: DeviceScope,
    ): InstalledModel {
        if (!scope.canControl) throw Refusal.NotAllowed
        val needle = normalise(phrase)
        if (needle.isEmpty()) throw Refusal.NoSuchModel(phrase)

        installed.firstOrNull { normalise(it.id) == needle }?.let { return it }

        fun pick(candidates: List<InstalledModel>): InstalledModel? = when {
            candidates.size == 1 -> candidates[0]
            candidates.size > 1 -> throw Refusal.Ambiguous(phrase, candidates.map { it.name })
            else -> null
        }

        pick(installed.filter { normalise(it.name) == needle })?.let { return it }
        // The bare id without its quantization — "gemma-3-12b-it" for
        // "gemma-3-12b-it@Q4_K_M" — which is how a person says it and how the catalog
        // prints it.
        pick(installed.filter { normalise(it.id.substringBefore('@')) == needle })?.let { return it }
        pick(installed.filter { normalise(it.name).startsWith(needle) })?.let { return it }
        pick(
            installed.filter {
                normalise(it.name).contains(needle) || normalise(it.id).contains(needle)
            },
        )?.let { return it }

        throw Refusal.NoSuchModel(phrase)
    }

    fun loadRequest(
        phrase: String,
        installed: List<InstalledModel>,
        scope: DeviceScope,
    ): LoadRequest =
        // An installed model carries its quantization in its id, so the request needs
        // nothing else: the Mac already knows what it has on disk.
        LoadRequest(modelID = resolveModel(phrase, installed, scope).id)

    /** "What is loaded on my Mac" — one sentence, spoken. */
    fun loadedSummary(status: Status, macName: String?): String {
        val mac = macName ?: "your Mac"
        val name = status.loadedModelName ?: status.loadedModelID
            ?: return "Nothing is loaded on $mac right now."
        val sentence = StringBuilder("$mac has $name loaded")
        status.contextLength?.takeIf { it > 0 }?.let {
            sentence.append(", with a ${contextPhrase(it)} context")
        }
        status.lastGenerationTokensPerSecond?.takeIf { it > 0 }?.let {
            sentence.append(", last answering at ${it.toInt()} tokens a second")
        }
        return "$sentence."
    }

    /** "65,536" spoken as "64K", because that is how anybody says it. */
    fun contextPhrase(tokens: Int): String {
        if (tokens < 1024) return "$tokens token"
        val thousands = tokens / 1024.0
        return if (thousands == kotlin.math.floor(thousands)) {
            "${thousands.toInt()}K"
        } else {
            String.format("%.1fK", thousands)
        }
    }

    /** Case, spacing and punctuation are not how a person distinguishes two models. */
    fun normalise(text: String): String = text.lowercase()
        .replace('_', '-')
        .replace(' ', '-')
        .trim('-', '.', ',', '!', '?', ' ')
}
